File: src/com/suushiemaniac/lang/json/value/JsonObject.java
package com.suushiemaniac.lang.json.value;

import com.suushiemaniac.lang.json.JSON;
import com.suushiemaniac.lang.json.util.StringUtils;
import com.suushiemaniac.lang.json.exception.JsonNotIterableException;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class JsonObject extends JsonElement {
    private SortedMap<String, JSON> members;

    public JsonObject(SortedMap<String, JSON> members) {
        this(null, members);
    }

    public JsonObject(JSON parent, SortedMap<String, JSON> members) {
        super(parent);

        for (Map.Entry<String, JSON> jsonTypeEntry : members.entrySet()) jsonTypeEntry.getValue().setParent(this);

        this.members = members;
    }

    @Override
    public JSON get(JSON keyIndex) {
    	try {
    		return this.members.get(keyIndex.stringValue());
		} catch (JsonNotIterableException e) {
			throw new JsonNotIterableException("Can't get child element by given JSON type. Expected: String, Found: " + keyIndex.type());
		}
    }

    @Override
    public JSON get(String key) {
        return this.members.get(key);
    }

    @Override
    public JSON get(int index) {
        return this.members.get(String.valueOf(index));
    }

    @Override
    public void set(JSON keyIndex, JSON value) {
        if (keyIndex instanceof JsonString) {
            value.setParent(this);
            this.members.put(keyIndex.stringValue(), value);
        }
    }

    @Override
    public void set(String key, JSON value) {
        value.setParent(this);
        this.members.put(key, value);
    }

    @Override
    public void set(int index, JSON value) {
        value.setParent(this);
        this.members.put(String.valueOf(index), value);
    }

    @Override
    public void set(JSON value) {
        if (value instanceof JsonObject) this.members = ((JsonObject) value).members;

        for (Map.Entry<String, JSON> tEntry : this.members.entrySet())
            tEntry.getValue().setParent(this);
    }

    @Override
    public void add(JSON value) {
        throw new JsonNotIterableException("Can't add child element without explicit key");
    }

    @Override
    public void add(JSON keyIndex, JSON value) {
        if (keyIndex instanceof JsonString) {
            value.setParent(this);
            this.members.put(keyIndex.stringValue(), value);
        }
    }

    @Override
    public void add(String key, JSON value) {
        value.setParent(this);
        this.members.put(key, value);
    }

    @Override
    public void add(int index, JSON value) {
		value.setParent(this);
		this.members.put(String.valueOf(index), value);
    }

    @Override
    public void remove(String key) {
        this.members.remove(key);
    }

    @Override
    public void remove(JSON value) {
        this.members.remove(value.stringValue());
    }

    @Override
    public void remove(int index) {
		this.members.remove("" + index);
    }

    @Override
    public JSON keyIndexOf(JSON content) {
        for (Map.Entry<String, JSON> entry : this.members.entrySet()) {
            if (entry.getValue().equals(content)) {
                return new JsonString(entry.getKey());
            }
        }

        return null;
    }

    @Override
    public void clear() {
        this.members.clear();
    }

    @Override
    public int size() {
        return this.members.size();
    }

    @Override
    public int deepSize() {
        int deepSize = 0;

        for (Map.Entry<String, JSON> tEntry : this.members.entrySet()) {
            deepSize += tEntry.getValue().deepSize();
        }

        return deepSize;
    }

    @Override
    public Iterator<JSON> iterator() {
        List<JSON> typeList = this.members.entrySet().stream().map(Map.Entry::getValue).collect(Collectors.toList());

        return typeList.iterator();
    }

    public String toString() {
        List<String> stringedList = new ArrayList<>();

        for (String key : this.members.keySet()) {
            stringedList.add(StringUtils.jsonWrap(key) + ":" + this.members.get(key).toString());
        }

        return "{" + String.join(",", stringedList) + "}";
    }

    @Override
    public boolean equals(Object other) {
        return other instanceof JsonObject && ((JsonObject) other).members.equals(this.members);
    }

    //TODO maybe simplify
    @Override
    public String toFormatString() {
        List<String> stringedList = new ArrayList<>();

        String tabbing = this.members.size() <= 1 ? "" : StringUtils.copy("\t", this.hierarchy() + 1);

        for (String key : this.members.keySet()) {
            JSON val = this.members.get(key);
            stringedList.add(tabbing + StringUtils.jsonWrap(key) + ": " + StringUtils.reduceTab(val.toFormatString()));
        }

        String openTabbing = StringUtils.copy("\t", this.hierarchy() - (this.parent() instanceof JsonElement && this.parent().size() <= 1 ? 1 : 0));
        String borderBreak = stringedList.size() <= 1 ? "" : "\n";
        String closingTabs = stringedList.size() <= 1 ? "" : openTabbing;
        return openTabbing + "{" + borderBreak + String.join(",\n", stringedList) + borderBreak + closingTabs + "}";
    }

    //TODO definitely simplify
    @Override
    public String toXMLString() {
        List<String> stringedList = new ArrayList<>();

        for (String key : this.members.keySet()) {
            String tagKey = StringUtils.jsonUnwrap(key);

            String openTag = "<" + tagKey + ">";
            String closeTag = "</" + StringUtils.jsonUnwrap(key) + ">";

            JSON child = this.members.get(key);

            if (child instanceof JsonObject) {
                child = new JsonObject(child.parent(), new TreeMap<>(((JsonObject) child).members));
                List<String> properties = new ArrayList<>();

                for (JSON childKey : child.keySet()) {
                    JSON childVal = child.get(childKey);

                    if (childVal instanceof JsonElement) continue;

                    properties.add(childKey.stringValue() + "=" + StringUtils.jsonWrap(childVal.toString()));
                    child.remove(childKey);
                }

                tagKey += " " + String.join(" ", properties);
            }

            if (child instanceof JsonArray && child.size() > 0) {
                openTag = closeTag = "";
            }

            String val = child.toXMLString();
            stringedList.add(val.length() > 0 ? openTag + val + closeTag : "<" + tagKey + "/>");
        }

        return String.join("\n", stringedList);
    }

    @Override
    public Object toNative() {
        return this.nativeMap();
    }

    @Override
    public Collection<JSON> collect() {
        return this.members.entrySet().stream().map(Map.Entry::getValue).collect(Collectors.toSet());
    }

    @Override
    public Set<JSON> keySet() {
        return this.members.keySet().stream().map(JsonString::new).collect(Collectors.toSet());
    }

    @Override
    public Map<String, Object> nativeMap() {
        SortedMap<String, Object> nativeMap = new TreeMap<>();

        for (String key : this.members.keySet())
        	nativeMap.put(key, this.members.get(key).toNative());

        return nativeMap;
    }

    @Override
    public List<Object> nativeList() {
		return this.nativeMap().entrySet().stream().map(Map.Entry::getValue).collect(Collectors.toList());
    }

    @Override
    public Set<Object> nativeSet() {
		return this.nativeMap().entrySet().stream().map(Map.Entry::getValue).collect(Collectors.toSet());
    }
}

File: src/main/kotlin/com/suushiemaniac/lang/json/value/JsonObject.kt
package com.suushiemaniac.lang.json.value

import com.suushiemaniac.lang.json.JSON
import com.suushiemaniac.lang.json.util.StringUtils.jsonWrap
import com.suushiemaniac.lang.json.util.StringUtils.jsonUnwrap
import com.suushiemaniac.lang.json.util.StringUtils.reduceTab
import com.suushiemaniac.lang.json.exception.JsonNotIterableException

import java.util.*

class JsonObject(parent: JSON?, private var members: SortedMap<String, JSON>) : JsonElement(parent) {
    constructor(members: SortedMap<String, JSON>) : this(null, members)

    init {
        for (value in members.values) value.parent = this
    }

    override fun get(keyIndex: JSON): JSON? {
        try {
            return this.members[keyIndex.stringValue()]
        } catch (e: JsonNotIterableException) {
            throw JsonNotIterableException("Can't get child element by given JSON type. Expected: String, Found: " + keyIndex.type())
        }

    }

    override fun get(key: String): JSON? {
        return this.members[key]
    }

    override fun get(index: Int): JSON? {
        return this.members[index.toString()]
    }

    override fun set(keyIndex: JSON, value: JSON) {
        if (keyIndex is JsonString) {
            value.parent = this
            this.members[keyIndex.stringValue()] = value
        }
    }

    override fun set(key: String, value: JSON) {
        value.parent = this
        this.members[key] = value
    }

    override fun set(index: Int, value: JSON) {
        value.parent = this
        this.members[index.toString()] = value
    }

    override fun set(value: JSON) {
        if (value is JsonObject) this.members = value.members

        for (memberVal in this.members.values)
            memberVal.parent = this
    }

    override fun add(value: JSON) {
        throw JsonNotIterableException("Can't add child element without explicit key")
    }

    override fun add(keyIndex: JSON, value: JSON) {
        this[keyIndex] = value
    }

    override fun add(key: String, value: JSON) {
        this[key] = value
    }

    override fun add(index: Int, value: JSON) {
        this[index.toString()] = value
    }

    override fun remove(key: String) {
        this.members.remove(key)
    }

    override fun remove(value: JSON) {
        this.members.remove(value.stringValue())
    }

    override fun remove(index: Int) {
        this.members.remove(index.toString())
    }

    override fun keyIndexOf(content: JSON): JSON {
        for ((key, value) in this.members) {
            if (value == content) {
                return JsonString(key)
            }
        }

        return JsonNull.INST
    }

    override fun clear() {
        this.members.clear()
    }

    override fun size(): Int {
        return this.members.size
    }

    override fun deepSize(): Int {
        return this.members.values.sumBy { it.deepSize() }
    }

    override fun iterator(): Iterator<JSON> {
        return this.members.values.iterator()
    }

    override fun toString(): String {
        return this.members.map { "${it.key.jsonWrap()}:${it.value}" }.joinToString(",", prefix = "{", postfix = "}")
    }

    override fun equals(other: Any?): Boolean {
        return other is JsonObject && other.members == this.members
    }

    override fun hashCode(): Int {
        return this.members.hashCode()
    }

    //TODO maybe simplify
    override fun toFormatString(): String {
        val tabbing = if (this.members.size <= 1) "" else "\t".repeat(this.hierarchy() + 1)

        val stringedList = this.members.map { "$tabbing${it.key.jsonWrap()}: ${it.value.toFormatString().reduceTab()}" }

        val openTabbing = "\t".repeat(this.hierarchy() - if (this.parent is JsonElement && this.parent!!.size() <= 1) 1 else 0)

        val borderBreak = if (stringedList.size <= 1) "" else "\n"
        val closingTabs = if (stringedList.size <= 1) "" else openTabbing

        return openTabbing + "{" + borderBreak + stringedList.joinToString(",\n") + borderBreak + closingTabs + "}"
    }

    //TODO definitely simplify
    override fun toXMLString(): String {
        val stringedList = ArrayList<String>()

        for ((key, child) in this.members) {
            var tagKey = key.jsonUnwrap()

            var openTag = "<$tagKey>"
            var closeTag = "</$tagKey>"

            if (child is JsonObject) {
                val properties = ArrayList<String>()

                for ((childKey, childVal) in child.members) {
                    if (childVal is JsonElement) continue

                    properties.add("$childKey=${childVal.toString().jsonWrap()}")
                    child.remove(childKey)
                }

                tagKey += properties.joinToString(" ", prefix = " ")
            }

            if (child is JsonArray && child.size() > 0) {
                closeTag = ""
                openTag = closeTag
            }

            val xmlVal = child.toXMLString()
            stringedList.add(if (xmlVal.isNotEmpty()) "$openTag$xmlVal$closeTag" else "<$tagKey/>")
        }

        return stringedList.joinToString("\n")
    }

    override fun toNative(): Any {
        return this.nativeMap()
    }

    override fun nativeKeySet(): Set<String> {
        return this.members.keys
    }

    override fun collect(): Collection<JSON> {
        return this.members.values.toSet()
    }

    override fun keySet(): Set<JSON> {
        return this.members.keys.map { JsonString(it) }.toSet()
    }

    override fun <T> nativeMap(mapping: (JSON) -> T): Map<String, T> {
        return this.members.mapValues { mapping(it.value) }
    }

    override fun <T> nativeList(mapping: (JSON) -> T): List<T> {
        return this.nativeMap(mapping).values.toList()
    }

    override fun <T> nativeSet(mapping: (JSON) -> T): Set<T> {
        return this.nativeMap(mapping).values.toSet()
    }
}